libcompat/fchmodat.c

```diff
@@ -0,0 +1,106 @@
+/*
+   Copyright (C) CFEngine AS
+
+   This file is part of CFEngine 3 - written and maintained by CFEngine AS.
+
+   This program is free software; you can redistribute it and/or modify it
+   under the terms of the GNU General Public License as published by the
+   Free Software Foundation; version 3.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
+
+  To the extent this program is licensed as part of the Enterprise
+  versions of CFEngine, the applicable Commercial Open Source License
+  (COSL) may apply to this file if you as a licensee so wish it. See
+  included file COSL.txt.
+*/
+
+#include <platform.h>
+#include <misc_lib.h>
+#include <logging.h>
+#include <generic_at.h>
+
+#include <fcntl.h>
+#include <grp.h>
+#include <errno.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#ifndef __MINGW32__
+
+typedef struct
+{
+    const char *pathname;
+    mode_t mode;
+    int dirfd;
+    int flags;
+} fchmodat_data;
+
+static int fchmodat_inner(void *generic_data)
+{
+    struct stat buf;
+	struct group *gr;
+    int ret = 0;
+	gid_t gid = 0, oldgid = 0;
+	uid_t olduid = 0;
+
+    fchmodat_data *data = generic_data;
+    if ((ret = fstatat(data->dirfd, data->pathname, &buf, data->flags)))
```

Comment:
Damn, I'm wrong again. This is tricky stuff to get right!

What we need to know is who owns the leaf, regardless of whether that is a link or something else. So always using ``lstat`` here is the correct approach. That means that noone can sneak in a link to a sensitive file, since the link would have to be owned by them, and hence would trigger the ``seteuid()`` below.

In addition, I think you should return error if ``flags`` contains ``AT_SYMLINK_NOFOLLOW``, since you cannot possibly honor it (we don't call this anywhere from our code, but it makes it future proof).